### ISSUE TITLE
Inconsistent type-casting behaviour

### ISSUE THREAD

**Issue by Wintermoose** [Microsoft/TypeScript#13699]
**TypeScript Version:**  2.1.5 (also 2.1.4)

**Code**

```ts
interface IFilter {
	render: string;
	color: number;
}

// 1a) this will compile
var a: IFilter = <IFilter>{};
// 1b) this will compile too
var b: IFilter = <IFilter>{color: 5};
// 2) this complains about missing render property
var c: IFilter = <IFilter>{color: 5, shape:3};
```

**Expected behavior:**
I am not entirely sure whether the behaviour 1 or 2 is the correct, one, but from the description of the type system I would expect 2) to be correct. Either way, I find it strange that the code behaves inconsistently in those 2 cases.

**Actual behavior:**
My expectation: all 3 casts fail, because of missing render property

**Motivation**
Just quick explanation of our usage of this: we have a method accepting array of objects (let's say filters) representing a declaration, the typecasts are used to make sure the individual objects contain all required properties (discriminate unions don't work here because new types can be mapped by the code).

**Comment by mhegazy** [comment]
Type assertion checks that either of the two types are assignable to each other (i.e. up-cast / down-cast). in this case, `IFilter` has a required `reader` property that is missing from `{color: 5, shape:3}` and `{color: 5, shape:3}` have a required `shape` property that is missing from `IFilter`. so the two types are unrelated.

Please see https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#416-type-assertions for more details.

You can assert to a common base type, as `<IFilter><{}>{color: 5, shape:3};` if you need to do the conversion.

**Comment by mhegazy** [comment]
For future references, we would recommend using StackOverFlow for similar questions.

**Comment by Wintermoose** [comment]
Ok sorry, I got sidetracked by the fact that the expression on the right side is a constant. When working with general expression the 2 directions of the "assignability" test make of course perfect sense. But when applied to a constant it seems somehow dubious (at least I cannot think if an example where such typecast isn't a disaster waiting to happen). Then again different of the type assertion for direct constants and dynamic expressions probably isn't desirable?
